msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md

```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
 
-## Known limitations
+## Obtaining consent
 
 Tenant admin consent: A simple way of [consenting on behalf of an organization as a tenant admin](/azure/active-directory/manage-apps/consent-and-permissions-overview#admin-consent) is by getting [consent from admin](/azure/active-directory/manage-apps/grant-admin-consent).
 
-You can ask for consent using the Auth API. Another approach for getting Graph scopes is to present a consent dialog using our existing [third party OAuth provider authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.
+Another approach for getting Graph scopes is to present a consent dialog using our existing [third party OAuth provider authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.
+
+You can also ask for consent using the TeamsJS [authentication](/javascript/api/@microsoft/teams-js/authentication) capability. 
 
 <details>
-<summary>To ask for additional consent using the Auth API, follow these steps:</summary>
+<summary>To obtain consent using the TeamsJS client library, follow these steps:</summary>
+
+> [!TIP]
+> The [Teams Personal Tab SSO Authentication](https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/main/samples/tab-personal-sso-quickstart/js/src/components/Tab.js#L64-L101) sample provides code demonstrating following steps.
 
 1. The token retrieved using `getAuthToken()` must be exchanged on the server-side using Azure AD [on-behalf-of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) to get access to those other Graph APIs. Ensure you use the v2 Graph endpoint for this exchange.
-2. If the exchange fails, Azure AD returns an invalid grant exception. It usually responds with one of the two error messages, `invalid_grant` or `interaction_required`.
-3. When the exchange fails, you must ask for consent. Use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog using [Silent authentication](~/concepts/authentication/auth-silent-aad.md).
-4. When asking for more consent from Azure AD, you must include `prompt=consent` in your [query-string-parameter](~/tabs/how-to/authentication/auth-silent-aad.md#get-the-user-context) to Azure AD, otherwise Azure AD wouldn't ask for other scopes.
+2. Often the first time you try to execute this token exchange for a user, Azure AD will refuse to exchange tokens because the user has not consented to giving your app permission to their data for this purpose. In these cases, your exchange will fail with either the `invalid_grant` or `interaction_required` error.
```

Comment:
Do we know why/when it will be one error vs the other? (I would @ mention Brian for this but for some reason he is now showing up in the dropdown for mentions... @iambmelt)

Reply:
Oh maybe it worked after all

Reply:
The semantics for `interaction_required` are specified by [OIDC](https://openid.net/specs/openid-connect-basic-1_0.html) and is AFAIK served from the `/authorization` endpoint.

`invalid_grant` is a broader error message (sometimes remediable through user interaction) that can occur at the `/token` endpoint (e.g., `auth_code` was invalid, malformed refresh token, bad request). I'm working off memory here and can dig deeper if you're interested

Reply:
In an OBO scenario should apps expect to possibly receive either of these errors? The main thing I want to ensure the docs are clear on is what developers should expect to have to handle.

Note: in the [OBO docs](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) it tells you to hit the [/token endpoint](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#middle-tier-access-token-request) but in the [error response example](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#error-response-example) on that same page, the example error given is `interaction_required` (even though the URL it said to use was `/token` and not `/authorization`).

Reply:
Ahhh there's some interesting behavior here, Conditional Access will solicit `interaction_required` from the token endpoint if a step-up is required, as these are user remediable. BUT, for reasons unapparent to me, needing consent is an `invalid_grant`, despite the error message telling you to go interactive.

```json
{
	"error": "invalid_grant",
	"error_description": "AADSTS65001: The user or administrator has not consented to use the application with ID 'XXXXX' named 'XXXXX'. Send an interactive authorization request for this user and resource.",
	"error_codes": [65001],
	"timestamp": "2023-06-16 15:51:31Z",
	"suberror": "consent_required"
}
```

Apps can absolutely receive both of these errors - the underlying reasons for them will be different. Some examples....

- `interaction_required` - MFA or device enrollment required. 
- `invalid_grant` - consent required or auth_code, assertion, or RT is expired, revoked, malformed, or absent

Reply:
Thanks, I'll add those examples to the explanation! Does "RT" stand for refresh token?

Reply:
I didn't write Brian's comment (obviously) but I believe that RT does indeed stand for refresh token.